public/app/features/dashboard/dashgrid/DashboardLibrary/CommunityDashboardMappingForm.tsx

```diff
@@ -0,0 +1,190 @@
+import { css } from '@emotion/css';
+import { useState } from 'react';
+
+import { DataSourceInstanceSettings, GrafanaTheme2 } from '@grafana/data';
+import { Trans, t } from '@grafana/i18n';
+import { getDataSourceSrv } from '@grafana/runtime';
+import { Stack, Text, Button, Alert, useStyles2, Field, Input } from '@grafana/ui';
+import { DataSourcePicker } from 'app/features/datasources/components/picker/DataSourcePicker';
+import { DashboardInput, DataSourceInput } from 'app/features/manage-dashboards/state/reducers';
+
+import { InputMapping, mapConstantInputs } from './utils/autoMapDatasources';
+
+interface Props {
+  dashboardName: string;
+  unmappedInputs: DataSourceInput[];
+  constantInputs: DashboardInput[];
+  existingMappings: InputMapping[];
+  onBack: () => void;
+  onPreview: (allMappings: InputMapping[]) => void;
+}
+
+export const CommunityDashboardMappingForm = ({
+  dashboardName,
+  unmappedInputs,
+  constantInputs,
+  existingMappings,
+  onBack,
+  onPreview,
+}: Props) => {
+  const styles = useStyles2(getStyles);
+  const [userDatasourceMappings, setUserDatasourceMappings] = useState<Record<string, DataSourceInstanceSettings>>({});
+  const [constantValues, setConstantValues] = useState<Record<string, string>>(() => {
+    // Initialize with default values from constantInputs
+    const initial: Record<string, string> = {};
+    constantInputs.forEach((input) => {
+      initial[input.name] = input.value;
+    });
+    return initial;
+  });
+
+  const handleDatasourceSelect = (inputName: string, datasource: DataSourceInstanceSettings) => {
+    setUserDatasourceMappings((prev) => ({
+      ...prev,
+      [inputName]: datasource,
+    }));
+  };
+
+  const handleConstantChange = (inputName: string, value: string) => {
+    setConstantValues((prev) => ({
+      ...prev,
+      [inputName]: value,
+    }));
+  };
+
+  const handlePreview = () => {
+    // Combine all mappings:
+    // 1. Existing auto-mapped datasources
+    // 2. User-selected datasources
+    // 3. Constant values (user-edited or defaults)
+
+    const userSelectedDatasources: InputMapping[] = unmappedInputs.map((input) => ({
```

Comment:
we're managing 2 different arrays here that have to be kind of synced. Why not to create this `userSelectedDatasources` state, initializing it with the unmappedInput `pluginId`s?
that state should be of type 
```
type UserDataSource = {
  pluginId: string; // or DataSourceInput, but we just use the pluginId
  datasource: DataSourceInstanceSettings;
}
```

that makes this easier

Reply:
it's something similar to what you did with `constantInputs`